### DISCUSSION TITLE
Vault tech overview

### DISCUSSION THREAD

**Discussion by stephenwf** [digirati-co-uk/iiif-canvas-panel#101] [General]
Vault is a javascript library with minimal dependencies for loading IIIF presentation resources. At a glance Vault will:

* Fetch or load IIIF JSON
* Upgrade Presentation 2 manifests to Presentation 3
* Normalise imported JSON
* Subscribe to changes at a granular level
* Track event listeners on data, instead of HTML elements
* Associated non-IIIF state alongside IIIF resources

## Getting started
For this walkthrough you can follow along with this [code sandbox](https://codesandbox.io/s/vault-vanilla-g5mzq?file=/index.html:1363-1368) or use one of the methods below:

<details>
<summary>Detailed installation instructions</summary>

#### NPM / Yarn projects

NPM Projects:
```
npm i @hyperion-framework/vault
```

Yarn projects:
```
yarn add @hyperion-framework/vault
```

and then import:
```js
import { Vault } from '@hyperion-framework/vault';
```
or with RequireJS / Node:
```js
const { Vault }  = require('@hyperion-framework/vault');
```

#### Script tag bundle

[Example code sandbox](https://codesandbox.io/s/vault-vanilla-bundle-dhtjw?file=/index.html)

```html
<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/@hyperion-framework/vault@1.1.0/dist/index.umd.js"></script>
<script>
    // This uses a global "HyperionVault" for the bundle.
    const vault = new HyperionVault.Vault();
</script>
```

#### Modern browser modules
[Example code sandbox](https://codesandbox.io/s/vault-vanilla-g5mzq?file=/index.html:1363-1368)

Include an import map in your `<head />`
```html
<script type="importmap">
  {
    "imports": {
      "@hyperion-framework/vault": "https://cdn.jsdelivr.net/npm/@hyperion-framework/vault/dist/index.es.js",
      "@hyperion-framework/store": "https://cdn.jsdelivr.net/npm/@hyperion-framework/store/dist/index.es.js",
      "@hyperion-framework/parser": "https://cdn.jsdelivr.net/npm/@hyperion-framework/parser/dist/index.es.js",
      "@hyperion-framework/presentation-2-parser": "https://cdn.jsdelivr.net/npm/@hyperion-framework/presentation-2-parser/dist/index.es.js",
      "@atlas-viewer/iiif-image-api": "https://cdn.jsdelivr.net/npm/@atlas-viewer/iiif-image-api/dist/index.es.js",
      "mitt": "https://cdn.jsdelivr.net/npm/mitt@3.0.0/dist/mitt.mjs",
      "redux": "https://cdn.jsdelivr.net/npm/redux@4.1.2/es/redux.mjs",
      "typesafe-actions": "https://cdn.jsdelivr.net/npm/typesafe-actions@5.1.0/dist/typesafe-actions.es.production.js"
    }
  }
</script>
```
and then create a new script tag:
```html
<script type="module">
    import { Vault } from "@hyperion-framework/vault";

    const vault = new Vault();
</script>
```

No bundler required :tada:

</details>

## Fetching JSON

The first thing you will do with Vault is load some IIIF into it.

Loading a manifest from a URL
```js
vault.loadManifest( ... ); // -> Promise<ManifestNormalized>
```

_You can load a Presentation 2 or Presentation 3 manifest into Vault. Any Presentation 2 manifests will be upgraded to Presentation 3. This allows you to create UI and components against the latest version of the specification and still support existing manifests._

When you load any resource into Vault it will do the following things to normalize the data:

<details>
<summary>Click to see all normalization steps</summary>

### Always arrays

Anything that can be optionally an array of values will always be an array. This is mostly a part of the IIIF specification now, but this will ensure that it can be relied on.
```json
{ "thumbnail": { "id": "https://example.org/image.jpg", "type": "Image" } }
```
<strong><div align="center">
⬇️  ⬇️  ⬇️ ⬇️ ⬇️  ⬇️  ⬇️ ⬇️  
</div></strong>

```json
{ "thumbnail": [{ "id": "https://example.org/image.jpg", "type": "Image" }] }
```

### Always empty

Any value that is omitted is given a default value of null or an empty array (for iterables).
```json
{ "id": "http://example.org/item" }
```

<strong><div align="center">
⬇️  ⬇️  ⬇️ ⬇️ ⬇️  ⬇️  ⬇️ ⬇️  
</div></strong>

```js
{ 
  "id": "http://example.org/item",
  "thumbnail": [],
  "label": null
}
```
(This covers _all_ properties, the above is an example of thumbnail and label only)

### Flattening
For all linked or structural properties we flatten resources.
```json
{
  "id": "https://example.org/manifest-1",
  "type": "Manifest",
  "label": {"none": ["Manifest 1"]},
  "items": [
    {
      "id": "https://example.org/manifest-1/c1",
      "type": "Canvas",
      "label": {"none": ["Canvas 1"]}
    }
  ]
}
```

<strong><div align="center">
⬇️  ⬇️  ⬇️ ⬇️ ⬇️  ⬇️  ⬇️ ⬇️  
</div></strong>

```json
{
  "Manifest": {
    "https://example.org/manifest-1": {
      "id": "https://example.org/manifest-1",
      "type": "Manifest",
      "label": {"none": ["Manifest 1"]},
      "items": [
        {
          "id": "https://example.org/manifest-1/c1",
          "type": "Canvas"
        }
      ]
    }
  },
  "Canvas": {
    "https://example.org/manifest-1/c1": {
      "id": "https://example.org/manifest-1/c1",
      "type": "Canvas",
      "label": {"none": ["Canvas 1"]}
    }
  }
}
```

</details>

## Getting data from the vault

When you have added data into the vault, you can retrieve them again with the `fromRef` method.

```js
const manifest = vault.fromRef({ id: 'http://example.org/manifest-1', type: 'Manifest' });
```

Due to the normalization steps, nested resources are only references when they come out of the Vault. If you wanted to load the list of canvases you can use either the `fromRef` for an individual resource, or `allFromRef`
```js
const manifest = vault.fromRef({ id: 'http://example.org/manifest-1', type: 'Manifest' });

// Returns only a reference to a canvas.
const firstCanvasRef = manifest.items[0]; // -> { id: 'http://example.org/canvas-1', type: 'Canvas' }

// Passing to fromRef resolves the whole canvas
const firstCanvas = vault.fromRef(manifest.items[0]); // -> { id: 'http://example.org/canvas-1', type: 'Canvas', items: ... , label: ... }

// Passing to allFromRef resolves all canvases
const allCanvases = vault.allFromRef(manifest.items); // -> [{ ... }, { ... }, { ... }]
```

Although this may feel like a trade-off of using this library, it would be feasible to create a wrapper around the Vault and offer your own utility functions on IIIF resources. **Vault is not a IIIF utility library, it's a fetching and storing library that utility libraries can be built on**.

You can also select directly from the Vault.
```js
const annotationPageId = 'https://example.org/annotation-page';
const annotationPage = vault.select(state => state.hyperion.entities.AnnotationPage[annotationPageId]);
```

## Dynamic data
Data can change for a few reasons in Vault. Some data, like annotation pages, could change once they are dereferenced. Others could change more often in an editing application. The state of the Vault may change quite often as new manifests are loaded. If you are building a dynamic application you may only want to know when the resource you are rendering has changed.

For that you can use subscriptions.

```js
const annotationPageId = 'https://example.org/annotation-page';
vault.subscribe(
  state => {
    // When this slice of the store changes...
    return state.hyperion.entities.AnnotationPage[annotationPageId],
  },
  annotationPage => {
    // This function will be called (and the slice of the store is passed in)
    renderMyAnnotationPage(annotationPage);
  }
);

// So here we've set up our subscriber. Now we can load it into Vault.
vault.load('https://example.org/annotation-page'); // This will trigger it straight away, before the data is loaded

// While it's loading, we will see the Identifier, but no annotations.
// When it has finished loading, our `renderMyAnnotationPage` will be called again with the full page.
```

Additionally you can make atomic edits to manifests, canvases or other resources in the Vault:
```js
const manifest = vault.fromRef({ ... });

// This works for any field.
vault.modifyEntityField(manifest, 'label', {  en: ['My new manifest label'] });
```
Anything subscribed to changes on that manifest will be notified / re-render. This can be used to drive anything from a full-manifest editing application to an annotation creation tool.

## Meta values
Vault stores the value of the IIIF, but sometimes applications may want to store and make available other state associated with IIIF resources. This data can be anything, and is enforced through convention only.

**Example**: Event listeners

Vault allows you to add event listeners onto IIIF resources. You can write the following:
```js
vault.addEventListener(manifest, 'onClick', (e) => {
  // Respond to click event.
})
```
On its own, this code will only make a record of the event listener. When you come to render a canvas elsewhere in your application, you can subscribe to the event listeners and use them.
```js
vault.subscribe(
    state => state.hyperion.meta[manifest.id],
    meta => {
       if (meta && meta.eventManager) {
          meta.eventManager.onClick // this is an array of event handlers that I can attach to my DOM elements.
       }
    }
)
```
On it's own this is not a good developer experience, but it is the low-level tools that can be abstracted. In the React abstraction of this library, this becomes a hook that can be used like this:
```jsx
function MyCanvas({ canvas }) {
  const canvasEvents = useResourceEvents(canvas);

  return <div {...canvasEvents}> ... </div>
}
```
And these will automatically change if you add new events to the Vault elsewhere in your application.

## Custom meta fields

In addition to using the build in meta values (event listeners, annotation pages, annotation styling) you can create custom ones for your own application.

To add meta to a resource, you can use: 
```js
vault.setMetaValue(
  [manifest.id, 'MyCustomStorage', 'myKey'],
  'myValue'
);

const state = {
  meta: {
    'https://example.org/manifest-1': {
      MyCustomStorage: { myKey: 'myValue' }
    }
  }
};
```
You can then subscribe to this like other other meta fields.
```js
function onUpdateMyKey(meta, manifest) {
  console.log(meta.MyCustomStorage);
 // ...
}

vault.subscribe(state => state.meta[manifest.id], onUpdateMyKey);
```

Set metadata value can also accept a function:
```js
function incrementCounterOnManifest(manifest) {
  vault.setMetaValue(
    [manifest.id, 'MyCustomStorage', 'myCounter'],
    (x = 0) => x + 1;
  );
}
```

**Comment by tomcrane** [comment]
* What's the difference between `vault.loadManifest` and `vault.load`? Are they the same operation? They _appear_ to be. 

* `fromRef` requires an object with `id` and `type` to match the vault stored object, e.g., 

```
        const manifest2 = vault.fromRef({ id: 'https://iiif.wellcomecollection.org/presentation/b1932795x', type: 'Manifest' });
        console.log(manifest2); // The manifest
        
        const manifest3 = vault.fromRef({ id: 'https://iiif.wellcomecollection.org/presentation/b1932795x' });
        console.log(manifest3); // just the object { id: 'https://iiif.wellcomecollection.org/presentation/b1932795x' }

        const manifest4 = vault.fromRef('https://iiif.wellcomecollection.org/presentation/b1932795x');
        console.log(manifest4); // just the string "https://iiif.wellcomecollection.org/presentation/b1932795x"
```

...is there a reason not to support just pulling out by ID?

**Reply by stephenwf** [reply]
* `loadManifest` and `load` are the same, if you are using VSCode though, you will get back a `Manifest` type with the former.
* Good idea, we do have a mapping for `id -> type` so not a big step to support that. Only again comes down to types. When you choose: `fromRef({ id: '...', type: 'Manifest' })` you get a manifest typed correctly in your IDE.

**Comment by tomcrane** [comment]
Does Vault have helpers for languageMaps? (If not, does it need them?)

One example I can think of is getting a string with fallbacks... 

`canvas.label.en[0] // might be null, maybe you'd be happy with the "none" string, or even a different language`

`vault.getLabel(canvas, "en", true)` e.g., falls back to "none", and if no "none", the first available string; still returns a string array

`vault.getSingleLabel(canvas, "en")` returns a string rather than an array. But is it the first string, or a concatenated string?

Fallback to "none" but don't fall back to other languages...
Fallback to other languages in this order...
Join multiple values using this separator

and so on...

**Reply by stephenwf** [reply]
This is a tough one. We build some utilities, but when you're building real-world apps - you want this to be baked into how your translations work too. 
In Madoc, for example, we have a `<LocaleString />` component that will pick the best language based on your current browser language, and handle ltr or rtl viewing direction.
https://github.com/digirati-co-uk/madoc-platform/blob/9a15ac0f2a2780fe645e92385ef332f480a0ff26/services/madoc-ts/src/frontend/shared/components/LocaleString.tsx

Which is used like this:
```jsx
<LocaleString>{canvas.label}</LocaleString>
```
and means you can largely ignore them with this component!

But this relies on the i18next framework, which provides languages and helpers specific to selecting the language.

**Reply by tomcrane** [reply]
Vault could provide this kind of thing... vault.getLabel(resource, locale [, locale, locale...])

(where locale can be a full string like "en-GB" or just the lang "en", or "none")

And then components / consumer code could do the work of obtaining that locale for the environment they are using vault in (e.g., a browser) but Vault still has the languageMap logic for reuse.
